### PR TITLE
feat(cdn):  use logrus to send worker and service log

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/cdn/cdn_log.go [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+func (s *Service) RunTcpLogServer(ctx context.Context) {
+	log.Info(ctx, "Starting tcp server %s:%d", s.Cfg.TCP.Addr, s.Cfg.TCP.Port)
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", s.Cfg.TCP.Addr, s.Cfg.TCP.Port))
+	if err != nil {
+		log.Fatalf("unable to start tcp log server: %v", err)
+	}
+
+	//Gracefully shutdown the tcp server
+	go func() {
```
Comment (fsamin) [inline engine/cdn/cdn_log.go:35]: sdk.Goroutine ?

File: engine/cdn/cdn_log.go [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+		select {
+		case <-ctx.Done():
+			log.Info(ctx, "CDN> Shutdown tcp log Server")
+			_ = listener.Close()
+		}
+	}()
+
+	go func() {
```
Comment (fsamin) [inline engine/cdn/cdn_log.go:43]: sdk.Goroutine ?

File: engine/cdn/cdn_log.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				log.Error(ctx, "unable to accept connection: %v", err)
+				return
+			}
+			go s.handleConnection(ctx, conn)
```
Comment (fsamin) [inline engine/cdn/cdn_log.go:50]: sdk.Goroutine ?

File: sdk/jws/jws.go [hunk 4]
```diff
@@ -63,6 +76,15 @@ func NewSigner(privateKey *rsa.PrivateKey) (jose.Signer, error) {
 	return jose.NewSigner(jose.SigningKey{Algorithm: jose.PS512, Key: privateKey}, nil)
 }
 
+// NewSigner instantiate a signer using HMAC using SHA-512 with the given private key.
```
Comment (yesnault) [inline sdk/jws/jws.go:79]:
```suggestion
// NewHMacSigner instantiates a signer using HMAC using SHA-512 with the given private key.
```

File: sdk/hatchery/service_log.go [hunk 5]
```diff
@@ -0,0 +1 @@
+package hatchery
```
Comment (yesnault) [inline sdk/hatchery/service_log.go:1]: empty file?

File: engine/hatchery/serve.go [hunk 6]
```diff
@@ -187,6 +190,46 @@ func (c *Common) getPanicDumpListHandler() service.Handler {
 	}
 }
 
+func (c *Common) InitServiceLogger() error {
+	tcpServer := c.Common.ServiceInstance.LogServer
+	var signer jose.Signer
+	if tcpServer.Addr != "" && tcpServer.Port != 0 {
```
Comment (yesnault) [inline engine/hatchery/serve.go:196]:
```suggestion
	if tcpServer.Addr != "" && tcpServer.Port > 0 {
```

File: engine/cmd_start.go [hunk 7]
```diff
@@ -79,9 +80,12 @@ This component operates CDS workflow repositories
 #### VCS
 This component operates CDS VCS connectivity
 
+#### CDN
+This component operates CDS CDN to handle storage
+
 Start all of this with a single command:
 
-	$ engine start [api] [hatchery:local] [hatchery:marathon] [hatchery:openstack] [hatchery:swarm] [hatchery:vsphere] [elasticsearch] [hooks] [vcs] [repositories] [migrate] [ui]
+	$ engine start [api] [hatchery:local] [hatchery:marathon] [hatchery:openstack] [hatchery:swarm] [hatchery:vsphere] [elasticsearch] [hooks] [vcs] [repositories] [cdn] [migrate] [ui]
```
Comment (yesnault) [inline engine/cmd_start.go:88]:
```suggestion
	$ engine start [api] [cdn] [hatchery:local] [hatchery:marathon] [hatchery:openstack] [hatchery:swarm] [hatchery:vsphere] [elasticsearch] [hooks] [vcs] [repositories] [migrate] [ui]
```

File: engine/cdn/cdn.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		case <-ctx.Done():
+			log.Info(ctx, "CDN> Shutdown HTTP Server")
+			_ = server.Shutdown(ctx)
+		}
+	}()
+
+	//Start the http server
+	log.Info(ctx, "CDN> Starting HTTP Server on port %d", s.Cfg.HTTP.Port)
+	if err := server.ListenAndServe(); err != nil {
+		log.Fatalf("CDN> Cannot start cds-cdn: %s", err)
```
Comment (yesnault) [inline engine/cdn/cdn.go:101]:
```suggestion
		log.Fatalf("CDN> Cannot start cds-cdn: %v", err)
```

File: engine/cmd_config.go [hunk 9]
```diff
@@ -203,6 +204,14 @@ var configCheckCmd = &cobra.Command{
 			}
 		}
 
+		if conf.CDN != nil && conf.CDN.API.HTTP.URL != "" {
```
Comment (fsamin) [inline engine/cmd_config.go:207]: for the moment i think conf.CDN.API.HTTP.URL is useless

File: engine/hatchery/serve.go [hunk 10]
```diff
@@ -187,6 +190,46 @@ func (c *Common) getPanicDumpListHandler() service.Handler {
...
+		if err != nil {
+			return sdk.WithStack(err)
+		}
+		signer, err = jws.NewSigner(c.Common.PrivateKey)
+		if err != nil {
+			return sdk.WithStack(err)
+		}
+		c.Signer = signer
+		c.ServiceLogger = logger
+	}
```
Comment (fsamin) [inline engine/hatchery/serve.go:207]: if we are unable to start a logger (because the conf is empty), we should raise an error

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13111.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/hatchery/serve.go (hunk 10 above)
Comment (sguiheux) [reply engine/hatchery/serve.go:207]: If conf is empty, it use the previous system. I want to keep this, to be able to rollback the feature by changing the configuration

File: engine/cmd_config.go (hunk 9 above)
Comment (sguiheux) [reply engine/cmd_config.go:207]: for the moment, all changes in this file is useless in fact, but we will need it for the next step.

File: engine/cdn/cdn_log.go (hunk 1 above)
Comment (sguiheux) [reply engine/cdn/cdn_log.go:35]: I don’t think for this particular usecase, but we can talk about it

File: engine/cdn/cdn_log.go (hunk 2 above)
Comment (sguiheux) [reply engine/cdn/cdn_log.go:43]: I don’t think for this particular usecase, but we can talk about it

File: engine/cdn/cdn_log.go (hunk 3 above)
Comment (sguiheux) [reply engine/cdn/cdn_log.go:50]: yep need it

File: engine/hatchery/serve.go (hunk 10 above)
Comment (sguiheux) [reply engine/hatchery/serve.go:207]: or enable it by configuration

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13112.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13118.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13120.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13124.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13126.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/cdn/cdn.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	"github.com/ovh/cds/engine/api/services"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/cdsclient"
+	"github.com/ovh/cds/sdk/log"
+)
+
+// New returns a new service
+func New() *Service {
+	s := new(Service)
+	/*
```
Comment (richardlt) [inline engine/cdn/cdn.go:17]: // TODO no router init for now because the service will run inside the api

File: engine/cdn/cdn_log.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+		if err != nil {
+			return err
+		}
+	} else {
+		workerData = cacheData.(sdk.Worker)
+	}
+	if err := jws.Verify(workerData.PrivateKey, sig.(string), &signature); err != nil {
+		return err
+	}
+	if workerData.JobRunID == nil || *workerData.JobRunID != signature.JobID {
```
Comment (richardlt) [inline engine/cdn/cdn_log.go:121]: Could be nice to add the jobrunid in worker database signature.

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13133.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
  * Docs ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

File: engine/cdn/cdn_router.go [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package cdn
+
+import (
+	"context"
+)
+
+func (s *Service) initRouter(ctx context.Context) {
+	//r := s.Router
```
Comment (richardlt) [inline engine/cdn/cdn_router.go:8]: Comment info // TODO not used for now because running inside api

File: engine/cdn/types.go [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+	"github.com/ovh/cds/engine/api/cache"
+	"github.com/ovh/cds/engine/service"
+	"github.com/ovh/cds/sdk"
+)
+
+// Service is the stuct representing a hooks µService
+type Service struct {
+	service.Common
+	Cfg Configuration
+	//Router *api.Router
```
Comment (richardlt) [inline engine/cdn/types.go:14]: Add a // TODO comment

File: sdk/log.go [hunk 15]
```diff
@@ -40,4 +40,5 @@ type ServiceLog struct {
 	ServiceRequirementID   int64      `json:"requirement_id" db:"-"`
 	ServiceRequirementName string     `json:"requirement_service_name" db:"requirement_service_name"`
 	Val                    string     `json:"val,omitempty" db:"value"`
+	WorkerName             string     `json:"worker_name"`
```
Comment (richardlt) [inline sdk/log.go:43]:
```suggestion
	WorkerName             string     `json:"worker_name" db:"-"`
```

**Comment by richardlt** [comment]
What about database signature for worker ? I think we should not ignore it but also this will require a migration to sign existing worker. Can we imagine another PR we the signature check.

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13193.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#13210.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Pull request merged**